File: js/src/instructions/registerFavoriteInstruction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerFavoriteInstruction = void 0;
const buffer_1 = require("buffer");
const web3_js_1 = require("@solana/web3.js");
const borsh_1 = require("borsh");
class registerFavoriteInstruction {
    constructor() {
        this.tag = 6;
    }
    serialize() {
        return (0, borsh_1.serialize)(registerFavoriteInstruction.schema, this);
    }
    getInstruction(programId, nameAccount, favouriteAccount, owner, systemProgram, optParent) {
        const data = buffer_1.Buffer.from(this.serialize());
        let keys = [];
        keys.push({
            pubkey: nameAccount,
            isSigner: false,
            isWritable: false,
        });
        keys.push({
            pubkey: favouriteAccount,
            isSigner: false,
            isWritable: true,
        });
        keys.push({
            pubkey: owner,
            isSigner: true,
            isWritable: true,
        });
        keys.push({
            pubkey: systemProgram,
            isSigner: false,
            isWritable: false,
        });
        if (!!optParent) {
            keys.push({
                pubkey: optParent,
                isSigner: false,
                isWritable: false,
            });
        }
        return new web3_js_1.TransactionInstruction({
            keys,
            programId,
            data,
        });
    }
}
exports.registerFavoriteInstruction = registerFavoriteInstruction;
registerFavoriteInstruction.schema = {
    struct: {
        tag: "u8",
    },
};

File: python/src/instructions/RegisterFavoriteInstruction.py
from typing import Optional

from borsh_construct import CStruct, U8
from solders.instruction import Instruction, AccountMeta
from solders.pubkey import Pubkey


class RegisterFavoriteInstruction:
    TAG = 6
    SCHEMA = CStruct(
        "tag" / U8,
    )

    @classmethod
    def serialize(cls) -> bytes:
        return cls.SCHEMA.build(dict(tag=cls.TAG))

    @classmethod
    def get_instruction(
        cls,
        program_id: Pubkey,
        name_account: Pubkey,
        favorite_account: Pubkey,
        owner: Pubkey,
        system_program: Pubkey,
        opt_parent: Optional[Pubkey] = None,
    ) -> Instruction:
        data = cls.serialize()

        accounts = [
            AccountMeta(
                pubkey=name_account,
                is_signer=False,
                is_writable=False,
            ),
            AccountMeta(
                pubkey=favorite_account,
                is_signer=False,
                is_writable=True,
            ),
            AccountMeta(
                pubkey=owner,
                is_signer=True,
                is_writable=True,
            ),
            AccountMeta(
                pubkey=system_program,
                is_signer=False,
                is_writable=False,
            ),
        ]

        if opt_parent:
            accounts.append(
                AccountMeta(
                    pubkey=opt_parent,
                    is_signer=False,
                    is_writable=False,
                )
            )

        return Instruction(
            program_id=program_id,
            accounts=accounts,
            data=data,
        )
